src/GafferSceneUI/Inspector.cpp

```diff
@@ -301,6 +310,303 @@ void Inspector::editScopeInputChanged( const Gaffer::Plug *plug )
+						return source;
+					}
+					else if( name == g_editWarningPropertyName )
+					{
+						return new StringData( editWarning );
+					}
+				}
+			}
+
+			if( currentHistory->predecessors.size() == 0 )
+			{
+				break;
+			}
+			currentHistory = currentHistory->predecessors[0].get();
+		}
+	}
+
+	return Path::property( name );
+}
+
+bool Inspector::HistoryPath::isLeaf( const Canceller *canceller) const
+{
+	return names().size() > 0;
+}
+
+PathPtr Inspector::HistoryPath::copy() const
+{
+	return new Inspector::HistoryPath( m_inspector, m_context, root(), const_cast<PathFilter *>( getFilter() ) );
+}
+
+void Inspector::HistoryPath::doChildren( std::vector<PathPtr> &children, const Canceller *canceller) const
+{
+	if( isLeaf() )
+	{
+		return;
+	}
+
+	std::string editWarning;
+
+	Context::Scope scope( m_context.get() );
+
+	SceneAlgo::History::ConstPtr history = m_inspector->history();
+	const SceneAlgo::History *currentHistory = history.get();
+
+	while( true )
+	{
+		Context::Scope currentScope( currentHistory->context.get() );
+
+		if( m_inspector->source( currentHistory, editWarning ) != nullptr )
+		{
+			children.push_back(
+				new Inspector::HistoryPath(
+					m_inspector,
+					m_context,
+					"/" + currentHistory->scene->relativeName( m_scriptNode.get() )
+				)
+			);
+		}
+
+		if( currentHistory->predecessors.size() == 0 )
+		{
+			break;
+		}
+
+		else if( currentHistory->predecessors.size() > 1 )
+		{
+			IECore::msg(
+				IECore::Msg::Warning,
+				"Inspector::HistoryPath",
+				"Branching histories are not supported, using first predecessor history only."
+			);
+		}
+
+		currentHistory = currentHistory->predecessors[0].get();
+
+	}
+	std::reverse( children.begin(), children.end() );
+}
+
+void Inspector::HistoryPath::pathChanged( Path *path )
+{
+	updateNodeNameChangedSignals();
+}
+
+void Inspector::HistoryPath::nodeNameChanged()
+{
+	emitPathChanged();
+}
+
+void Inspector::HistoryPath::inspectorDirtied( Inspector *inspector )
+{
+	emitPathChanged();
+}
+
+void Inspector::HistoryPath::updateNodeNameChangedSignals()
+{
+	using NodeSet = std::unordered_set<Gaffer::GraphComponent *>;
+
+	NodeSet newNodes;
+	std::vector<PathPtr> newChildren;
+
+	children( newChildren );
+
+	if( newChildren.size() == 0 )
+	{
+		return;
+	}
+
+	// Collect the end plug and its ancestor nodes, all of which are part of the path name.
+	for(
+		std::vector<PathPtr>::const_iterator it = newChildren.begin(), eIt = newChildren.end();
+		it != eIt;
+		++it
+	)
+	{
+		GraphComponent *node = m_scriptNode->descendant( (*it)->names()[0] );
+		while( node != m_scriptNode.get() )
+		{
+			newNodes.insert( node );
+			node = node->parent();
+		}
+	}
+
+	// Remove connections that no longer exist
+
+	for(
+		NodeMap::const_iterator it = m_nodeNameConnections.begin(), eIt = m_nodeNameConnections.end();
+		it != eIt;
+		++it
+	)
+	{
+		if( std::find( newNodes.begin(), newNodes.end(), it->first ) == newNodes.end() )
+		{
+			m_nodeNameConnections.erase( it );
```

Comment:
Technically, this invalidates `it`, so I don't think we're allowed to do `++it` on it afterwards. Instead I think we need to use the return value, which gives us the next iterator :

```
for( auto it = c.begin(); c != c.end(); /* empty */ )
{
    if( ... )
    {
        it = c.erase( it );
    }
    else
    {
        ++it;
    }
}
```

But now we've replaced the insanely slow `boost::signals`, I'd be pretty tempted to just clear all connections unconditionally and then rebuild the ones we want.

Reply:
Fixed in 3ec55b33222cc6fe8e8aad459b50bdda6efbc638